internal/workflow/activities/unarchive.go

```diff
@@ -0,0 +1,178 @@
+	// the archive should be excluded from the destination directory.
+	StripTopLevelDir bool
+}
+
+type UnarchiveResult struct {
+	// DestPath is the path to the extracted archive contents.
+	DestPath string
+
+	// IsDir is true if DestPath is a directory.
+	IsDir bool
+}
+
+func NewUnarchiveActivity(logger logr.Logger) *UnarchiveActivity {
+	return &UnarchiveActivity{logger: logger}
+}
+
+// Execute attempts to unarchive the contents of SourcePath to a temporary
+// directory. If SourcePath points to a directory or a non-archive file then the
+// path is returned, unaltered, as DestPath.
+func (a *UnarchiveActivity) Execute(ctx context.Context, params *UnarchiveParams) (*UnarchiveResult, error) {
+	a.logger.V(1).Info("Executing UnarchiveActivity", "Path", params.SourcePath)
+
+	fi, err := os.Stat(params.SourcePath)
+	if err != nil {
+		return nil, fmt.Errorf("unarchive: stat: %v", err)
+	}
+	if fi.IsDir() {
+		a.logger.V(2).Info("Unarchive: skipping directory")
+		return &UnarchiveResult{DestPath: params.SourcePath, IsDir: true}, nil
+	}
+
+	u, err := unarchiver(params.SourcePath)
+	if err != nil {
+		return nil, fmt.Errorf("unarchive: unarchiver: %v", err)
+	}
+	if u == nil {
+		// Couldn't find an unarchiver, so this is probably a regular file.
+		// Return the source path unaltered, and IsDir false.
+		a.logger.V(2).Info("Unarchive: not an archive, skipping")
+		return &UnarchiveResult{DestPath: params.SourcePath}, nil
+	}
+
+	dest := filepath.Join(filepath.Dir(params.SourcePath), "extract")
+	if err := u.Unarchive(params.SourcePath, dest); err != nil {
+		return nil, fmt.Errorf("unarchive: unarchive: %v", err)
+	}
+
+	if params.StripTopLevelDir {
+		if err = stripDirContainer(dest); err != nil {
+			return nil, fmt.Errorf("unarchive: strip top-level dir: %v", err)
+		}
+	}
+
+	if err := fsutil.SetFileModes(dest, ModeDir, ModeFile); err != nil {
+		return nil, fmt.Errorf("unarchive: %v", err)
+	}
+
+	if err := os.Remove(params.SourcePath); err != nil {
+		a.logger.V(1).Info("Unarchive: couldn't delete source archive: %v", err)
+	}
+
+	return &UnarchiveResult{DestPath: dest, IsDir: true}, err
+}
+
+// Unarchiver returns the unarchiver suited for the archival format.
+func unarchiver(filename string) (archiver.Unarchiver, error) {
+	if iface, err := archiver.ByExtension(filename); err == nil {
+		if u, ok := iface.(archiver.Unarchiver); ok {
+			return u, nil
+		}
+	}
+
+	file, err := os.Open(filepath.Clean(filename))
+	if err != nil {
+		return nil, fmt.Errorf("open file: %v", err)
+	}
+	defer file.Close()
+
+	if u, err := archiver.ByHeader(file); err == nil {
+		return u, nil
+	}
+
+	return nil, nil
+}
+
+// stripDirContainer strips the top-level directory of a transfer.
+func stripDirContainer(dir string) error {
+	tld, err := topLevelDir(dir)
+	if err != nil {
+		return fmt.Errorf("get top-level dir: %v", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "enduro-")
+	if err != nil {
+		return fmt.Errorf("make temp dir: %v", err)
```

Comment:
not sure the error "make temp dir" adds information to the error here. might as well just return the error.

Reply:
I don't actually know what the error output from `os.MkdirTemp()` looks like.  My current philosophy on Go error messages is to add some context to make it easy to find the caller (e.g. Oh, this error is in "unarchive: stripDirContainer: make temp dir: [error message]").   I notice I haven't included the "stipDirContainer" context in this case though, which is probably helpful to add.

Reply:
https://github.com/uber-go/guide/blob/master/style.md#error-wrapping reads that the underlying error can be returned as-is if there is no additional context that can be added, but I think I tend to always add some context just like you did because I can't never remember which functions in the standard library come with enough context. I think I'd be ok with both approaches, the worst that could happen is to have a little repetition here and there as opposed to no context at all.

> I notice I haven't included the "stipDirContainer" context in this case though, which is probably helpful to add.

Yeah good point, I feel like we haven't done this enough. I could be a nice practice to adopt.


On this topic, while browsing [pkgsite](https://github.com/golang/pkgsite) the other day I saw a pretty nifty [pattern using named return values and defer](https://github.com/golang/pkgsite/blob/ea43129276ed1c7e85557321bd2af0504fbc4a7f/internal/auth/auth.go#L34-L55) to wrap the error with the name of the function.

```
func Header(ctx context.Context, jsonCreds []byte, useExp bool) (_ string, err error) {
	defer derrors.Wrap(&err, "auth.Header(jsonCreds, %t)", useExp)
	// [...]
}
```

Sadly it only works in combination with named return parameters (try [playground](https://go.dev/play/p/wqeTti2btoT)) which are avoided by a lot of people, e.g. see [nonamedreturns](https://github.com/firefart/nonamedreturns supported by golangci-lint. It's in the [spec](https://go.dev/ref/spec#Defer_statements) but I guess they can be misleading?

Reply:
That's nice @sevein, thanks!

Make sure you use and **return** `err` inside `// [...]`. I think the example you shared will miss the extra wrapping done in lines [46](https://github.com/golang/pkgsite/blob/ea43129276ed1c7e85557321bd2af0504fbc4a7f/internal/auth/auth.go#L46) and [51](https://github.com/golang/pkgsite/blob/ea43129276ed1c7e85557321bd2af0504fbc4a7f/internal/auth/auth.go#L51) since that's not returning the actual `err` like in line [42](https://github.com/golang/pkgsite/blob/ea43129276ed1c7e85557321bd2af0504fbc4a7f/internal/auth/auth.go#L42).

I use this pattern to be able to join errors from defer functions and return them with the original one.

Reply:
E.g.: https://github.com/artefactual-sdps/preprocessing-sfa/blob/main/internal/workflow/preprocessing.go#L49

Reply:
I did actually include the error message "stripDir:" context at https://github.com/artefactual-sdps/enduro/blob/dev/issue-869-unarchive-activity/internal/workflow/activities/unarchive.go#L77

Reply:
Thanks @sevein for the background chat. It actually seems to work without returning the named return: https://go.dev/play/p/m8j5Gr0ZHDZ

Reply:
> https://github.com/uber-go/guide/blob/master/style.md#error-wrapping reads that the underlying error can be returned as-is if there is no additional context that can be added, but I think I tend to always add some context just like you did because I can't never remember which functions in the standard library come with enough context (they're definitely not consistent). I think I'd be ok with both approaches, the worst that could happen is to have a little repetition, which is better than no context.
> 
> > I notice I haven't included the "stipDirContainer" context in this case though, which is probably helpful to add.
> 
> Yeah good point, I feel like we haven't done this enough. I could be a nice practice to adopt.
> 
> On this topic, while browsing [pkgsite](https://github.com/golang/pkgsite) the other day I saw a pretty nifty [pattern using named return values and defer](https://github.com/golang/pkgsite/blob/ea43129276ed1c7e85557321bd2af0504fbc4a7f/internal/auth/auth.go#L34-L55) to wrap the error with the name of the function.
> 
> ```go
> func Header(ctx context.Context, jsonCreds []byte, useExp bool) (_ string, err error) {
> 	defer derrors.Wrap(&err, "auth.Header(jsonCreds, %t)", useExp)
> 	// [...]
> }
> ```
> 
> Sadly it only works in combination with named return parameters (try [playground](https://go.dev/play/p/ME004ReW6J2)) which are avoided by a lot of people, e.g. see [nonamedreturns](https://github.com/firefart/nonamedreturns) supported by golangci-lint. It's in the [spec](https://go.dev/ref/spec#Defer_statements) but I guess they can be misleading? Idk.

I believe the major reason named returns are not encouraged is because  they allow naked return with those values which can get harry pretty quickly. For example if you have a named return here you can do something like 

```go
func add(a, b int) (sum int) {
    sum = a + b
    return
} 
// this return the sum of a and b
```

https://go.dev/play/p/22eUUbi0vsI for the example with @jraddaoui will return two EOF errors to the console with a new line.